<h1 align="center"> Python Básico</h1>

## Sumário

- Instalando o Python (Anaconda)
- IDEs
- Hello World!
- Matemática Básica
- Strings
- Listas 
- if elif else
- for loops
- Funções
- Tuplas
- Conjuntos
- Dicionários


## Por que Python?

Python é uma linguagem aberta (Free) - https://github.com/python/cpython. Muitas pessoas ao redor do mundo estão trabalhando continuamente para melhorar a linguagem. Principais empresas de tecnologias usam o Python como uma de suas linguagens (google, facebook, dropbox, yelp, etc.) 

Nature recomenda [(link)](http://www.nature.com/news/programming-pick-up-python-1.16833).

Está entre as linguagens mais amadas e mais desejadas de acordo com a [pesquisa StackOverflow de 2019](https://insights.stackoverflow.com/survey/2019).

**"Python is the most wanted language for the third year in a row, meaning that developers who do not yet use it say they want to learn it."**

## Por que IPython/Jupyter?

Jupyter Notebook é uma aplicação popular que permite o usuário o inserir textos (formato markdown/html), rodar código e visualizar resultados em um único ambiente web. 

Isso permite a criação de documentos 'vivos'.

Exemplos de livros desenvolvidos com o Jupyter Notebook:
 - [Probabilistic Programming & Bayesian Methods for Hackers](https://github.com/CamDavidsonPilon/Probabilistic-Programming-and-Bayesian-Methods-for-Hackers)
 - [Python Data Science Handbook](https://github.com/jakevdp/PythonDataScienceHandbook)
 - [CFD Python, a.k.a. the 12 steps to Navier-Stokes](https://github.com/barbagroup/CFDPython)

[Clique aqui](https://jupyter-notebook.readthedocs.io/en/stable/examples/Notebook/Notebook%20Basics.html) para mais detalhes sobre como utilizar o Jupyter Notebook.



### Instalando Anaconda
A distribuição Anaconda possui os principais pacotes para uso ciêntífico, simplicando o processo de instalação e evitando conflitos de requisitos.

Operating System | Blog Post | Youtube Video
--- | --- | ---
Mac | [Install Anaconda on Mac](https://medium.com/@GalarnykMichael/install-python-on-mac-anaconda-ccd9f2014072#.lvhw2gt3k "Install Anaconda on Mac") | [Youtube Video](https://www.youtube.com/watch?v=B6d5LrA8bNE "Youtube Video")
Ubuntu | [Install Anaconda on Ubuntu](https://medium.com/@GalarnykMichael/install-python-on-ubuntu-anaconda-65623042cb5a#.4kwsp0wjl "Install Anaconda on Ubuntu") | [Youtube Video](https://www.youtube.com/watch?v=jo4RMiM-ihs "Youtube Video")
Windows | [Install Anaconda on Windows](https://medium.com/@GalarnykMichael/install-python-on-windows-anaconda-c63c7c3d1444#.66f7y3whf "Install Anaconda on Windows") | [Youtube Video](https://www.youtube.com/watch?v=dgjEUcccRwM "Youtube Video")

## IDEs
O Python pode ser também utilizado com diferentes IDEs (Integrated Development Environment).

- PyCharm - https://www.jetbrains.com/pycharm/
- Visual Studio Code - https://code.visualstudio.com/
- Atom - https://atom.io/
- Sublime - https://www.sublimetext.com/
- notepad++ - https://notepad-plus-plus.org/

## Tipos Básicos

O principais tipos básicos de dados no Python são números (ints e floats), booleanos e strings.

### Hello World!
Caracteres entre aspas

In [64]:
# This line is a comment
print('Hello World!')

Hello World!


Exemplo em outras linguagens:

### C:

```C
#include <stdio.h>

int main(void)
{
    printf("hello, world\n");
}
```

### C++:

```C++
#include <iostream>

int main()
{
    std::cout << "Hello, world!\n";
    return 0;
}
``` 

### JAVA:
```java
class HelloWorldApp {
    public static void main(String[] args) {
        System.out.println("Hello World!"); // Prints the string to the console.
    }
}
```

## Matemática Básica
Os números inteiros (ex. 2, 4, 20) são do tipo int, aqueles com parte fracionária (ex. 5.0, 1.6) são do tipo float. Veremos mais sobre tipos numéricos posteriormente neste tutorial.

In [65]:
# Add two ints
2 + 2

4

In [66]:
50 - 5*6

20

In [67]:
(50 - 5*6) / 4

5.0

In [68]:
8 / 5  # division always returns a floating point number

1.6

In [69]:
17 / 3  # classic division returns a float

5.666666666666667

In [70]:
17 // 3  # floor division discards the fractional part

5

In [71]:
17 % 3  # # the % operator returns the remainder of the division

2

In [72]:
5 * 3 + 2  # result * divisor + remainder

17

Com Python, podemos usar o operador ** para calcular potências 1:

In [73]:
5 ** 2 # 5 squared

25

In [74]:
2 ** 7 # 2 to the power of 7

128

O sinal de igual ('=') é usado para atribuir um valor a uma variável. Depois de uma atribuição, nenhum resultado é exibido antes do próximo prompt:

In [75]:
width = 20

In [76]:
height = 5 * 9

In [77]:
width * height

900

Se uma variável não é “definida” (não tem um valor atribuído), tentar utilizá-la gerará um erro:

In [78]:
# Not defined
#n

Além de int e float, o Python suporta outros tipos de números, tais como Decimal e Fraction. O Python também possui suporte nativo a números complexos, e usa os sufixos j ou J para indicar a parte imaginária (por exemplo, 3+5j).

In [79]:
(0 + 1j)**2

(-1+0j)

Além de números, Python também pode manipular strings (sequências de caracteres), que podem ser expressas de diversas formas. Elas podem ser delimitadas por aspas simples ('...') ou duplas ("...") e teremos o mesmo resultado 2. \ pode ser usada para escapar aspas:

In [80]:
'spam eggs'  # single quotes

'spam eggs'

In [81]:
'doesn\'t'  # use \' to escape the single quote...

"doesn't"

In [82]:
"doesn't"  # ...or use double quotes instead

"doesn't"

In [83]:
'"Yes," they said.'

'"Yes," they said.'

In [84]:
"\"Yes,\" they said."

'"Yes," they said.'

In [85]:
'"Isn\'t," they said.'

'"Isn\'t," they said.'

In [86]:
print('"Isn\'t," they said.')

"Isn't," they said.


In [87]:
print('C:\some\name')  # here \n means newline!

C:\some
ame


In [88]:
print(r'C:\some\name')  # note the r before the quote

C:\some\name


As strings literais podem abranger várias linhas. Uma maneira é usar as aspas triplas: """...""" ou '''...'''. O fim das linhas é incluído automaticamente na string, mas é possível evitar isso adicionando uma \ no final. O seguinte exemplo:

In [89]:
print("""\
Usage: thingy [OPTIONS]
     -h                        Display this usage message
     -H hostname               Hostname to connect to
""")

Usage: thingy [OPTIONS]
     -h                        Display this usage message
     -H hostname               Hostname to connect to



In [90]:
print("""\
Usage: thingy [OPTIONS]
     -h                        Display this usage message
     -H hostname               Hostname to connect to
""")

Usage: thingy [OPTIONS]
     -h                        Display this usage message
     -H hostname               Hostname to connect to



Strings podem ser concatenadas (coladas) com o operador +, e repetidas com *:

In [91]:
# 3 times 'un', followed by 'ium'
3 * 'un' + 'ium'
'unununium'

'unununium'

In [92]:
'a' ' b'

'a b'

As strings podem ser indexadas (subscritas), com o primeiro caractere como índice 0. Não existe um tipo específico para caracteres; um caractere é simplesmente uma string cujo tamanho é 1:

In [93]:
word = 'Python'

In [94]:
word[0]  # character in position 0

'P'

In [95]:
word[5]  # character in position 5

'n'

Índices também podem ser números negativos para iniciar a contagem pela direita:

In [96]:
word[-1]  # last character

'n'

In [97]:
word[-2]  # second-last character

'o'

In [98]:
word[-6]

'P'

Além da indexação, o fatiamento também é permitido. Embora a indexação seja usada para obter caracteres individuais, fatiar permite que você obtenha substring:

In [99]:
word[0:2]  # characters from position 0 (included) to 2 (excluded)

'Py'

In [100]:
word[2:5]  # characters from position 2 (included) to 5 (excluded)

'tho'

Os índices do fatiamento possuem padrões úteis; um primeiro índice omitido padrão é zero, um segundo índice omitido é por padrão o tamanho da string sendo fatiada:

In [101]:
word[:2]   # character from the beginning to position 2 (excluded)

'Py'

In [102]:
word[4:]   # characters from position 4 (included) to the end

'on'

In [103]:
word[-2:]  # characters from the second-last (included) to the end

'on'

<div style="float: center;"><img style="float: center;" src="http://www.nltk.org/images/string-slicing.png" align=center /></div>

A decisão por índice 0 e exclusão do segundo índice no fatiamento no Python é baseada [nesse artigo](https://www.cs.utexas.edu/users/EWD/ewd08xx/EWD831.PDF) de Edsger W. Dijkstra.

A função embutida `len()` devolve o comprimento de uma string:

In [104]:
s = 'supercalifragilisticexpialidocious'
len(s)

34

Python inclui diversas estruturas de dados compostas, usadas para agrupar outros valores. A mais versátil é **list (lista), que pode ser escrita como uma lista de valores (itens) separados por vírgula, entre colchetes**. Os valores contidos na lista não precisam ser todos do mesmo tipo.

In [105]:
squares = [1, 4, 9, 16, 25]
squares

[1, 4, 9, 16, 25]

Como strings (e todos os tipos de sequence nativos), listas pode ser indexados e fatiados:

In [106]:
squares[0]  # indexing returns the item

1

In [107]:
squares[-1]

25

In [108]:
squares[-3:]  # slicing returns a new list

[9, 16, 25]

In [109]:
squares.append(36)
squares

[1, 4, 9, 16, 25, 36]

In [110]:
letters = ['a', 'b', 'c', 'd']
letters

['a', 'b', 'c', 'd']

In [111]:
len(letters)

4

## if statement

Provavelmente o mais conhecido comando de controle de fluxo é o if. Por exemplo:

In [112]:
x = 42
if x < 0:
    # Python usa indentação para os blocos
    x = 0
    print('Negative changed to zero')
elif x == 0:
    print('Zero')
elif x == 1:
    print('Single')
else:
    print('More')

More


Pode haver zero ou mais partes elif, e a parte else é opcional. A palavra-chave ‘elif’ é uma abreviação para ‘else if’, e é útil para evitar indentação excessiva. Uma sequência if … elif … elif … substitui os comandos switch ou case, encontrados em outras linguagens.

Comparison Operator | Function
--- | --- 
< | less than
<= | less than or equal to
> | greater than
>= | greater than or equal to
== | equal
!= | not equal

Logical Operator | Description
--- | ---
and | If both the operands are True then condition becomes True.
or | If any of the two operands are True then condition becomes True. 
not | Used to reverse the logical (not False becomes True, not True becomes False)

In [113]:
num = 4
num > 0 and num  < 15

True

### Exercício:
1. Atribuir um valor int a uma variável num. 
2. Escrever um bloco if-else que faça o print "O seu int é par" ou "O seu int é impar".

Dica: `%`

In [114]:
num = 4
if num % 2 == 0:
    print("O int é par")
else: 
    print("O int é impar")

O int é par


### Exercício:
1. Atribuir uma string a uma variável word.
2. Escrever um bloco if-else que informe se a palavra tem mais de 2 letras 'a' ou menos.

Dica: Após construir a string, digite o nome da variável seguida por um ponto (`word.`) e 
pressione tab para ver os métodos disponíveis.

O `tab` também ajuda na completação das palavras.

Se quiser instruções de como usar a função, posicione o cursor dentro dos parênteses e 
pressione `shift + tab`.

Por exemplo, se digitarmos word.replace() e pressionarmos `shift + tab` dentro dos parênteses 
veremos a seguinte instrução:

```
Signature: word.replace(old, new, count=-1, /)
Docstring:
Return a copy with all occurrences of substring old replaced by new.

  count
    Maximum number of occurrences to replace.
    -1 (the default value) means replace all occurrences.

If the optional argument count is given, only the first count occurrences are
replaced.
Type:      builtin_function_or_method
```



In [115]:
## 
word = 'palavra'
if word.count('a') > 2:
    print('Mais de 2 letras a')
elif word.count('a') == 2:
    print('2 letras a')
else: 
    print('Menos de 2 letras a')   

2 letras a


## For Loops

O comando for em Python é um pouco diferente do que costuma ser em C ou Pascal. Ao invés de sempre iterar sobre uma progressão aritmética de números (como no Pascal), ou permitir ao usuário definir o passo de iteração e a condição de parada (como C), o comando for do Python itera sobre os itens de qualquer sequência (seja uma lista ou uma string), na ordem que aparecem na sequência. Por exemplo:

In [5]:
# Measure some strings:
words = ['cat', 'window', 'palavra']
for w in words:
    print(w, len(w))

cat 3
window 6
palavra 7


Se você precisa iterar sobre sequências numéricas, a função embutida range() é a resposta. Ela gera progressões aritméticas:

In [117]:
for i in range(5):
    print(i)

0
1
2
3
4


### Exercício
Construa um código para printar a posição da letra 'a' em cada palavra da lista 'words'.

In [6]:
for word in words:
    for i in range(len(word)):
        if word[i] == 'a':
            print(f'Palavra {word} com letra a na posição {i}')

Palavra cat com letra a na posição 1
Palavra palavra com letra a na posição 1
Palavra palavra com letra a na posição 3
Palavra palavra com letra a na posição 6


In [118]:
for word in words:
    for i, letter in enumerate(word):
        if letter == 'a':
            print(f'Palavra {word} com letra a na posição {i}')

Palavra cat com letra a na posição 1
Palavra defenestrate com letra a na posição 9


Dizemos quem um objeto é **iterável**, isso é, candidato a ser alvo de uma função ou construção que espera alguma coisa capaz de retornar sucessivamente seus elementos um de cada vez. Nós vimos que o comando for é um exemplo de construção, enquanto que um exemplo de função que recebe um iterável é sum():

In [119]:
sum(range(4))  # 0 + 1 + 2 + 3

6

In [120]:
list(range(4))

[0, 1, 2, 3]

O laço de repetição while executa enquanto a condição permanece verdadeira.

In [121]:
# Fibonacci series:
# the sum of two elements defines the next
a, b = 0, 1
while a < 10:
    print(a)
    a, b = b, a+b

0
1
1
2
3
5
8


# Funções

A palavra reservada def inicia a definição de uma função. Ela deve ser seguida do nome da função e da lista de parâmetros formais entre parênteses. Os comandos que formam o corpo da função começam na linha seguinte e devem ser indentados.

In [122]:
def fib(n):    # write Fibonacci series up to n
    """Print a Fibonacci series up to n."""
    a, b = 0, 1
    while a < n:
        print(a, end=' ')
        a, b = b, a+b
    print()

# Now call the function we just defined:
fib(2000)

0 1 1 2 3 5 8 13 21 34 55 89 144 233 377 610 987 1597 


A instrução `return` finaliza a execução e retorna um valor da função.

In [123]:
def fib2(n):  # return Fibonacci series up to n
    """Return a list containing the Fibonacci series up to n."""
    result = []
    a, b = 0, 1
    while a < n:
        result.append(a)    # see below
        a, b = b, a+b
    return result

f100 = fib2(100)    # call it
f100                # write the result

[0, 1, 1, 2, 3, 5, 8, 13, 21, 34, 55, 89]

Funções também podem ser chamadas usando argumentos nomeados da forma chave=valor. Por exemplo, a função a seguir:

In [124]:
def foo(a, b='b', c='c'): 
    print(a, b, c)

In [125]:
foo('a')
foo(a='a')
foo(a='a', b='b1')
foo(a='a', c='c1')

a b c
a b c
a b1 c
a b c1


In [126]:
foo(b='b1', c='c1')

TypeError: foo() missing 1 required positional argument: 'a'

In [127]:
foo(b='b1', 'c1')

SyntaxError: positional argument follows keyword argument (<ipython-input-127-491215a84917>, line 1)

In [128]:
foo('a', a='a1')

TypeError: foo() got multiple values for argument 'a'

# Exercício

Escreva uma função `right_justify` que tenha como argumento uma string s e que print a string com espaços suficientes para que a última letra da string esteja na coluna 60 do display:

```python
>>> right_justify('monty')
>>> right_justify('python')
                                                        monty
                                                       python
```

In [1]:
def right_justify(s):
    pass


# Exercício

Escreva uma função que retorne o valor absoluto de um número.

In [174]:
def absolute_value(x):
    pass

# Tuplas

Uma tupla consiste em uma sequência de valores separados por vírgulas, por exemplo:

12345

(12345, 54321, 'hello!')

In [131]:
# Tuples may be nested:


((12345, 54321, 'hello!'), (1, 2, 3, 4, 5))

In [132]:
# Tuples are immutable:

TypeError: 'tuple' object does not support item assignment

In [133]:
# but they can contain mutable objects:


([1, 2, 3], [3, 2, 1])

A instrução `t = 12345, 54321, 'hello!'` é um exemplo de empacotamento de tupla: os valores `12345`, `54321` e `'bom dia!'` são empacotados em uma tupla. A operação inversa também é possível:

In [134]:
# Desempacotamento


In [5]:
# split


In [13]:
# Exemplo de função que retorna uma tupla


# Conjuntos

Python também inclui um tipo de dados para conjuntos, chamado set. Um conjunto é uma coleção desordenada de elementos, sem elementos repetidos. Usos comuns para conjuntos incluem a verificação eficiente da existência de objetos e a eliminação de itens duplicados. Conjuntos também suportam operações matemáticas como união, interseção, diferença e diferença simétrica.

Chaves ou a função `set()` podem ser usados para criar conjuntos. Note: para criar um conjunto vazio você precisa usar `set()`, não `{}`; este último cria um dicionário vazio, uma estrutura de dados que discutiremos na próxima seção.

Uma pequena demonstração:

In [135]:
# show that duplicates have been removed

{'banana', 'apple', 'orange', 'pear'}


In [136]:
# fast membership testing

True

False

In [138]:
# Demonstrate set operations on unique letters from two words

# unique letters in a


{'a', 'b', 'c', 'd', 'r'}

In [139]:
# letters in a but not in b

{'b', 'd', 'r'}

In [141]:
# letters in both a and b

{'a', 'c'}

# Exercício
Crie uma lista de telefones em que cada elemento é uma tupla com o seguinte formato:
`('Nome', Tel)`

Use os dados seguintes para preencher a lista:

Nome | Tel 
--- | --- 
jack | 4098
john | 4139

Em seguida escreva uma função `find_tel` que receba uma string com o nome da pessoa e retorne o número do telefone.

In [35]:
# lista de tels
tel = 

def find_tel(s):
    pass


# Dicionários

Outra estrutura de dados muito útil embutida em Python é o dicionário, cujo tipo é `dict` (ver [Mapping Types — dict](https://docs.python.org/pt-br/3/library/stdtypes.html#typesmapping)). Dicionários são também chamados de “memória associativa” ou “vetor associativo” em outras linguagens. Diferente de sequências que são indexadas por inteiros, dicionários são indexados por chaves (keys), que podem ser de qualquer tipo imutável (como strings e inteiros).

Um bom modelo mental é imaginar um dicionário como **um conjunto não-ordenado de pares chave:valor**, onde as chaves são únicas em uma dada instância do dicionário. Dicionários são delimitados por chaves: `{}`, e contém uma lista de pares chave:valor separada por vírgulas. Dessa forma também será exibido o conteúdo de um dicionário no console do Python. O dicionário vazio é `{}`.

**Lista é um mapeamento por índices, dicionário é um mapeamento por chaves.**

{}

{}

In [44]:
# Create dict


{'jack': 4098, 'john': 4139, 'guido': 4127}

In [45]:
# Get value for specific key

4098

In [46]:
# Remove and add new key, value pair

{'jack': 4098, 'guido': 4127, 'irv': 4127}

True

False

In [49]:
# Iterate on keys


jack
guido
irv


In [28]:
# Iterante on key, value pairs


jack 4098
guido 4127
irv 4127


# Exercício

Crie uma função `histogram` que conte quantas vezes cada letra/caractere aparece no texto deste exercício.

Dica:
1. Use 3 aspas ('''texto''') para criar a string;
2. Crie um dicionário vazio;
3. Construa um for loop para iterar em cada letra da string;
4. No dicionário acumule cada ocorrência da letra na string correspondente.

In [1]:
def histogram(s):
    pass